parsec/core/fs/workspacefs/entry_transactions.py

```diff
@@ -198,17 +199,30 @@ async def entry_missing_data(
                     total_size += chunk.raw_size
 
         block_ids = [chunk.id for chunk in blocks]
-        missing_size = 0
 
-        local_chunk_ids = await self.local_storage.block_storage.get_local_chunk_ids(block_ids)
+        local_blocks = []
+        remote_blocks = []
+        local_and_remote_blocks = []
+
+        # To avoid concurrency problems local storage is called first
+        local_and_remote_chunk_ids = await self.local_storage.block_storage.get_local_chunk_ids(
+            block_ids
+        )
+        local_and_remote_chunk_ids_set = set(local_and_remote_chunk_ids)
+
+        local_chunk_ids = await self.local_storage.chunk_storage.get_local_chunk_ids(block_ids)
         local_chunk_ids_set = set(local_chunk_ids)
-        for chunk in blocks:
-            if chunk.id not in local_chunk_ids_set:
-                assert chunk.access is not None
-                missing_blocks.append(chunk.access)
-                missing_size += chunk.raw_size
 
-        return missing_size, total_size, missing_blocks
+        for chunk in blocks:
+            assert chunk.is_block
+            if chunk.id in local_and_remote_chunk_ids_set:
+                local_and_remote_blocks.append(chunk.access)
+            elif chunk.id in local_chunk_ids_set:
+                local_blocks.append(chunk.access)
+            else:
+                remote_blocks.append(chunk.access)
+
+        return local_and_remote_blocks, local_blocks, remote_blocks
```

Comment:
So `total_size` is no longer returned? I was thinking it might be interesting to return:
- `file_size`: `manifest.size`
- `proper_blocks_size`: `sum(chunk.raw_size for chunks in manifest.blocks for chunk in chunks if chunk.is_block())`
- `pending_chunks_size`: `sum(chunk.raw_size for chunks in manifest.blocks for chunk in chunks if not chunk.is_block())`

Reply:
This makes 6 values to return so maybe a `namedtuple` would be nice to avoid confusion.